### ISSUE TITLE
Move LLG related codes to a super class

### ISSUE THREAD

**Issue by davidcortesortuno** [computationalmodelling/fidimag#66]
Many of the micromagnetic driver classes inherit from the `llg.py` file, which is itself a driver. For good practice we should move most of the LLG code into a super class that contains the initialisation and saving functions. We can use the `Sim` class for this since currently it does not contain any relevant code and we avoid creating a new file that could possibly break most of the tests. We can also leave the `RHS` functions for Sundials as an abstract function on `Sim`, so every driver sets its own way of integrating the LLG equation.

We already started discussing this with Weiwei for the atomistic part of the code, and there is a comment that is worth considering when trying to make this change:
 https://github.com/computationalmodelling/fidimag/commit/48e9eb1c23580f3b94740f9d513ea360e91c03d7#commitcomment-17922207

The easiest is to start with the micromagnetic code.

**Comment by ww1g11** [comment]
@davidcortesortuno in some sense our LLG class is something like a Simulation class, in which most of code contribute to the initialisation, logical operations (such as run_until and relax methods) and data saving. That's why our sim.py is almost empty.  Actually, only the RHS function belongs to the real "LLG" code, so I haven't created a new class for it (please see the sundials_rhs method). Therefore, it's probably not necessary to separate llg.py into two classes, what do you think?

**Comment by ww1g11** [comment]
@davidcortesortuno  Sorry that I haven't noticed you already started it, yes, your new structure is very impressive. In your new structure, LLG class is inherited from the MicroDriver, so do you have detailed plan on which parts will be placed in the MicroDriver class? I am also quite interesting in this topic, so maybe we should draw a figure to show the relation between Sim, MicroDriver, LLG and other LLG-based class?

**Comment by davidcortesortuno** [comment]
Hi @ww1g11 , yes I started changing some parts, we were discussing this yesterday. I kind of understand what you did the first time to keep most of the stuff in the LLG class, the problem is that it has started to become complex as you added extra terms to the LLG, like STT. It is a mess now that, for example, when we load `Sim`, then if you choose `STT LLG` it inherits from `LLG`. Our idea is to move most of the common properties to the Simulation class (saving options, process data, mesh info, etc) and then we have a class that ONLY handles the integration (I called `micro_driver`). Then, the LLG, STT LLG, etc. only have the `sundials_rhs` and `relax` functions, that have to do with their own integration. Currently, the problem is to figure out how are going to share the `spin` , `alpha`, `pins`, etc. arrays to this `driver` class, since those arrays are also important in the `Sim` class, we can maybe just reference them but we have to think about it. The other thing is that restructuring this will allow us to document every function that was in the LLG class, so a user can see what every part of the code is doing. The task is not easy as you see, since there is a lot of already written code, however if we manage doing this, we assure that in the future, adding new features will be easier, since adding a new driver will be possible only adding a couple of equations to be integrated.

**Comment by ww1g11** [comment]
Hi @davidcortesortuno, I still think you will make things a bit compilcated than the existing code, :-) and that's why this is not an easy thing, ;-) 

>Currently, the problem is to figure out how are going to share the spin , alpha, pins, etc. arrays to this >driver class, since those arrays are also important in the Sim class,
This is one reason that why I want to use one class rather than two individual classes. In theory, the entry point is the Sim class, so all the important data and methods should be called for users starting with a Simulation instance. For example, the following usage is allowed:
   ...
   sim.alpha = 0.01
   ...
   sim.relax(...)
   ...

This means that if you put arrays or methods in LLG class, you still have to rewrite them in Sim class, so to avoid rewritting them again, it's better to put them in Sim class? I am not sure. But for some not so common data such as a_J and P in  [LLG_STT_CPP](https://github.com/computationalmodelling/fidimag/blob/ENH-066-Sim_Refactor/fidimag/micro/llg_stt_cpp.py) class should be stored in stt_llg_cpp class.

>The other thing is that restructuring this will allow us to document every function that was in the LLG >class, so a user can see what every part of the code is doing. 
yes, indeed, this is good chance to write the documention.


>the problem is that it has started to become complex as you added extra terms to the LLG, like STT. 
by the way, it seem right now it's pretty easy to add new equations, for example, in [LLG_STT_CPP](https://github.com/computationalmodelling/fidimag/blob/ENH-066-Sim_Refactor/fidimag/micro/llg_stt_cpp.py) class we only need to rewrite `sundials_rhs` method. 

But anyway, it's a good idea to try new structure, so please keep on going and we will see how it evolve, :-)

**Comment by ww1g11** [comment]
Hi @davidcortesortuno , I have drawn a UML figure to show your idea that making use of supper LLG class, please see the attached picture. 
<img width="808" alt="fidimag3" src="https://cloud.githubusercontent.com/assets/12461209/17081659/7e792282-5193-11e6-8eaa-ca177a8dd773.png">
In the figure, as you can see, we have a supper class named Sim, but our atomistic.LLG, SLLG, Monte_Carlo and micro.LLG will inherit from the Sim class. Furthermore, LLG-based class such as spin transfer torques will inherit from LLG class.

This design is different from the code you have pushed recently, I know you are trying to sperate the code into different classes according to their properties, for example, you think LLG, and other LLG based classes are 'drivers', so they should be put into an individual class. As you already mentioned, you guys still debate on which place should be put the common data such as spin. So I am thinking why should we just use inheritance to avoid this kind of problem? although the LLG and Sim classes are acutally not sperated in this way. 

If we use the structure in the attached figure, we do not have to change too much code, for example, LLG_STT,  LLG_STT_CPP,  and Baryakhtar are already inherited from LLG class, so we even don't have to touch them when we start to use the new Sim class, :-)

**Comment by davidcortesortuno** [comment]
Thanks for the diagram, looks great. I don't know if it would be a good idea to leave Sim in the common folder. I think separating the simulations for now it's ok, in order to distinguish what we are simulating. Are the SLLG and MonteCarlo only defined in the atomistic framework?

I don't understand quite well why you think that the Sim class would grow as we add new equations. The Sim class would only be to keep common functions. The drivers will have their own functions on their own file, and you would only to create a new file if you want to add a new equation, which I think it's better, rather than maybe change something in the LLG class if you want a new common functionality. I also think that, intuitively, the LLG is just an extra driver, and most of the things are in the Sim class. Is this how Finmag works?


As for the current state of Fidimag, I used a package called `pyreverse` that comes with `pylint` to generate an UML diagram of the code, to have a better idea of the code structure :

![classes_no_name](https://cloud.githubusercontent.com/assets/12200541/17098689/a85fc164-525b-11e6-96d9-b46386de574d.png)

**Comment by ww1g11** [comment]
Hi @davidcortesortuno The UML diagram is very cool, thank you. It seems that there are some duplications between the two LLG classes, so a super Class between them would reduce the duplications significantly. This supper class don't have to be the Sim class, but if we use Sim class it would be easy for us. Alternatively,  a new Sim class, as the unique entry point, will behave as an interface and thus we have to consider all the situations related to whole fidimag.  As an example, you can generate the same UML diagram for finmag, and you will see how complicated it is, :-) Anyway, choose the one you like and go ahead, enjoy refactoring!

**Comment by davidcortesortuno** [comment]
Hi, as an update, I have started refactoring things in a branch: https://github.com/computationalmodelling/fidimag/tree/ENH-066-Sim_Refactor

Now we have a `SimBase` and a `DriverBase` classes that contain common functionalities between the atomistic and micromagnetic functions. The LLG , LLG_STT, etc. now inherit from the corresponding driver class. For example, for the micromagnetic case we have:

```
DriverBase <-- MicroDriver <-- (LLG, LLG_STT, SLLG, LLG_STT_CPP)
```

The `Microdriver` in this case, is referenced in the micromagnetic `Simulation` class. This cleans up a lot of code and I have get rid of a lot of duplication.

I have also removed the dependence on the `SaveVTK` class that seems useless. I left it for now since the NEB classes (and MonteCarlo I think) still use it, but we can remove this class in the future.

A diagram of the current state of the code is here:

![classes_no_name](https://cloud.githubusercontent.com/assets/12200541/17514556/2ae5949e-5e2b-11e6-95a6-7cf79d2dd43b.png)


Most of the tests are passing, I think some the notebooks are failing. In general, nothing has changed much. The most notorious change is that we know have to relax with `sim.driver.run_until` or `sim.driver.relax` (I made a link to the driver to still use `sim.relax`). The gyromagnetic ratio `gamma` now has to change from `sim.driver.gamma` as well as specific parameters in the drivers, for example in the LLG_STT we now have to do `sim.driver.beta` to change it.

What is needed now are more tests, since we have split a lot of the code and should be more easy to test. Then, we can later write a documentation on how to extend Fidimag easily, just adding stuff in the right places
